### ISSUE TITLE
Create `except-object` exception to `arrow-body-style` rule, enforcing explicit return of object literals from arrow functions.

### ISSUE THREAD

**Issue by duncanbeevers** [eslint/eslint#5936]
The `arrow-body-style` rule enforces braceless function bodies when the function consists of a single return statement.

When constructing arrow functions which return object literals, the return value must be wrapped in parentheses to force its evaluation as an expression.

The proposed new option enforces a block body with an explicit return value for object literals returned from arrow functions.

Valid
```js
var foo = () => { return; { bar: 0 }; };
var foo = () => 0;
```

Invalid
```js
var foo = () => ({ bar: 0 });
```

**Comment by nzakas** [comment]
I'm not sure I'm understanding this correctly.  You're proposing to add an option that requires the use of `return` if you want to return an object literal, is that correct?

**Comment by duncanbeevers** [comment]
Yes, exactly.

**Comment by BYK** [comment]
I think this is a fair request.

**Comment by nzakas** [comment]
@byk are you championing this change?

My feedback:

* It's unclear to me if this is really a third option, equal to "always" and "as-needed". It seems like it could actually apply in addition to " as-needed". This needs some discussion. 
* I don't think "except-object" is a good name. We generally try to stay away from using "except" because it's unclear. We tend to use "allow", " skip", etc., words that have clearer meaning.

**Comment by duncanbeevers** [comment]
Yes, the naming for this is a little awkward. "except-object" is a refinement of "as-needed", not a distinct third option.

Conceivable, this could live as its own eslint plugin (`parenthesized-object-literals` or something) but in practice arrow functions are they only place I've seen them appear, as well as the only place I want to enforce this.

**Comment by BYK** [comment]
Don't want to over commit so holding on for championing for now in case someone else can do it. If we can't find anyone else in a week I'll take it on. Sounds good? 

Also agree that this needs more discussion and consideration before moving forward with the code.

**Comment by BYK** [comment]
@eslint/eslint-team - anyone else interested in this rule? Either championing or contributing to the discussion?

**Comment by mikesherov** [comment]
I vote that this should be a third option as its not truly needed.

**Comment by nzakas** [comment]
@byk it's been a week, are you championing? :)

Also, agree with @mikesherov that this would be a third option.

**Comment by BYK** [comment]
@nzakas - okay you got me, championing.

**Comment by nzakas** [comment]
Okay, we just need three :+1:s to move forward.

I'll give you one: :+1:

**Comment by mikesherov** [comment]
:+1:

**Comment by alberto** [comment]
👍  @BYK I could also take this

**Comment by BYK** [comment]
@alberto I'm just getting back to speed so if you have free time go for it. Otherwise I'll get to this eventually :)

**Comment by alberto** [comment]
Cool

**Comment by alberto** [comment]
I added the `needs bikeshedding` because I think we still need to settle on the option name (sorry for not being explicit). Is that tag appropriate for this?

**Comment by alberto** [comment]
I'm having trouble finding a name for a standalone option :( Any ideas?

**Comment by duncanbeevers** [comment]
What do you think of a third argument which refines the existing "as-needed" setting?

```js
`"arrow-body-style": [2, "as-needed", "skip-object-literal"]
```

**Comment by alberto** [comment]
That would be a refinement for `as-needed`, not a third option, which is what @nzakas and @mikesherov proposed. Also I am not sure the terminology `skip` (or `allow`) is clear in this case.

**Comment by nzakas** [comment]
We could do `allowObjectLiteralBody` with a default value of `true` and let people switch to `false` to force the use of curlies. 

@byk as champion, can you drive this discussion to a conclusion?

**Comment by duncanbeevers** [comment]
@byk @alberto 

What do you think of using an extension similar to the `avoidQuotes` option used for the [`object-shorthand`](http://eslint.org/docs/rules/object-shorthand) rule?

```js
"arrow-body-style": [2, "as-needed", { avoidParens: true }]
```

**Comment by nzakas** [comment]
I don't think `avoidParens` is specific enough. I would expect that option to also flag `a => (B)`

**Comment by mysticatea** [comment]
`avoidWrappedObjectLiterals`?

**Comment by BYK** [comment]
I'm +1 to `allowObjectLiteralBody: true` suggestion for now. If we get 2 more 👍s I'd say we go with that.

**Comment by alberto** [comment]
I'm sorry but I'm still not sure if you intend that as a new option or as a refinement for `as-needed`. @nzakas @BYK

**Comment by platinumazure** [comment]
Added "needs bikeshedding" label since we're still hashing a few things out.

**Comment by BYK** [comment]
I think this should be an option for `as-needed` mode.

**Comment by alberto** [comment]
Thanks, that makes more sense to me. I was bit confused. Should we use `requireForObjectLiteral` instead?. My concern is `allow` doesn't convey not using braces is discouraged in that case.

**Comment by BYK** [comment]
@alberto - I'm not really sure actually. `requireForObjectLiteral` implies "require 'arrow function body' for object literal" which is an option to make the rule more strict. AFAIK we tend to have options that relax the original rule and `allowObjectLiteralBody` fits into that.

What do you think?

**Comment by alberto** [comment]
Ok, I think I see your point.

**Comment by ilyavolodin** [comment]
It looks like there are two PRs for this issue: #5934 and #6216 Should we close one of them?

**Comment by BYK** [comment]
I'd close #5934 since it was created before the issue was accepted (hence before all these discussions). That said since @duncanbeevers is a new collaborator, there's a case to get his first patch in too. What do you think @ilyavolodin and @alberto.

**Comment by mikesherov** [comment]
I think we should favor new contributions here.

**Comment by ilyavolodin** [comment]
Agree with @mikesherov @alberto Would you mind guiding @duncanbeevers with the necessary updates to his PR?

**Comment by alberto** [comment]
oh, sorry, I didn't see that PR since it wasn't linked to the issue.

@duncanbeevers are you still interested in implementing this?

**Comment by duncanbeevers** [comment]
I appreciate being favored for contribution, but I @alberto's #6216 is already done and ready for inclusion. I can't imagine implementing it any differently than they did, so I'd prefer #6216 to #5934

**Comment by BYK** [comment]
@duncanbeevers - thanks! Alright then, I closed #5934 and I think #6216 is ready to merge.

**Comment by alberto** [comment]
ok, @duncanbeevers. Thanks for your contribution anyway!
